### ISSUE TITLE
Semiglobal split-read alignment

### ISSUE THREAD

**Issue by bkehr** [seqan/seqan#1679]
I would like to compute a split read alignment with free end gaps in the read sequences:

```
reference: aaaaaAAAAAAxxxxxxxxxxxxxxBBBBBBBbbbbbb
read:      -----AAAAAA--------------BBBBBBB------
```

More precisely, I would like to compute the split position where the read jumps from one reference to another:

```
referenceA: aaaaaAAAAAAAAxxxxxxxxxxxx
referenceB: xxxxxxxxxxxxxBBBBBBbbbbbbbb
read:            AAAAAAAABBBBBB

Alignment A: aaaaaAAAAAAAA
             -----AAAAAAAA

Alignment B: BBBBBBbbbbbbbb
             BBBBBB--------
```

The `splitAlignment()` function currently penalizes the gaps in `a` and `b` columns of the example. It doesn't allow to specify an `AlignConfig` object for free end gaps.

However, when I do the following two changes in the the `align_split` module, it seems to work:

1. align_split/align_split_interface.h: Replace line 300 by 

    `typedef FreeEndGaps_<False, True, True, True> TFreeEndGaps;`
2. align_split/dp_scout_split.h: Comment line 121.

Can someone (@rrahn?) who understands the align module's entire logic confirm that this will work?
It would be nice if the `splitAlignment()` function had an `AlignConfig` parameter.

Here is a small test program:

```
//#include <seqan/align_split.h>
#include "align_split.h"

using namespace seqan;

int main()
{
	DnaString refLeft  = "TTTTTTTTTTTTGAGCCGATTTTTTTT";
	DnaString refRight = "TTTTTTTTTTTTTTTTGGACCGTTTTTTTTTTTTTTTTTTTTTTt";
	DnaString read     = "GAGCCGA" "GGACCG";

	Gaps<DnaString> refGapsLeft;
	Gaps<DnaString> refGapsRight;
	Gaps<DnaString> readGapsLeft;
	Gaps<DnaString> readGapsRight;

	setSource(refGapsLeft, refLeft);
	setSource(refGapsRight, refRight);
	setSource(readGapsLeft, read);
	setSource(readGapsRight, read);

	Score<int> scoring(1, -3, -4, -5);

	int splitScore = splitAlignment(readGapsLeft, refGapsLeft, readGapsRight, refGapsRight, scoring);

	std::cout << "refLeft:  " << refGapsLeft << "\n";
	std::cout << "readLeft: " << readGapsLeft << "\n\n";

	std::cout << "refRight:  " << refGapsRight << "\n";
	std::cout << "readRight: " << readGapsRight << "\n\n";

	std::cout << "Split score = " << splitScore;
}
```

Current output:
```
refLeft:  TTTTTTTTTTTTG
readLeft: GAGCCGAGGACCG

refRight:  
readRight: 

Split score = -35
```
Output when applying the above two changes:
```
refLeft:  TTTTTTTTTTTTGAGCCGA
readLeft: ------------GAGCCGA

refRight:  GGACCGTTTTTTTTTTTTTTTTTTTTTTT
readRight: GGACCG-----------------------

Split score = 13
```

**Comment by rrahn** [comment]
Hey, I'll have a look at this.

**Comment by rrahn** [comment]
Mmmh, this needs more thoughts I guess.
You are actually looking for a local alignment, is that right?
The current split align code only works as global alignment, so just removing the line 121 in dp_scout_split.h would invalidate the current split alignment.
But I added an interface using align config, to specify which gaps should be penalised. 
But you example from above would still not work:

```
refLeft:  TTTTTTTTTTTTGAGCCGATTTTTTTT-------
readLeft: ---------------------------GAGCCGA

refRight:  ------TTTTTTTTTTTTTTTTGGACCGTTTTTTTTTTTTTTTTTTTTTTT
readRight: GGACCG---------------------------------------------

Split score = 13
```

**Comment by bkehr** [comment]
Yes and no:
Yes because I think the traceback should start in a cell of the DP matrix that is not necessarily in the last row or column.
No because I want to have a penalty for gaps in the reference sequence.

Actually, I think the traceback cannot be computed before finding the split position as it should start in the cell of the split position. Since the split DP scout only stores one value per column, it would probably be the maximum in that column. Does that make sense?

**Comment by rrahn** [comment]
Ok, I managed it to get the code working as you expected it in your test case. 
The trick was, that you need a semi-global alignment as suggested, but needed to tell the underlying algorithm to only trace the part that is enabled.
By default, the optimum can come from either end (read or config), while in the semi-global version you can only come from one end. You achieved this little detail by commenting this one line as suggested.
But this would have break the default case. So I added some metafunction dispatching, to allow specifying whether the last row or column should be traced.
That solves the use case from above. 

@bkehr Can you please verify that this works for you? Here is the PR  #1685.

**Comment by bkehr** [comment]
Splendid, it works for me when adding `AlignConfig<false, true, true, false> config;` to the test case above.

**Comment by bkehr** [comment]
Hi again,

I found a case where it doesn't work and I need your help in finding out what goes wrong:

```
    DnaString read     = "AGGGGTTTCGCCATGTTGGCCTGGCTGGTCTCGAATTCCTGACCTCAGGTGATCCACCCGCCCTGGCCTCCCAAAGTGCTGGGATTACAGGCCTGAGCCGCTGCACCCGGCCCAGAGGCATTTATTTAAAATGTGCAATAAAAGTGTTCTG";
    DnaString refLeft  = "AGGCGCACGCCACCACCAAGCCCGGCTAAAGTTTTGTGTTTTTAGTATAGATGGGGTTTCGCCATGTTGGCCTGGCTGGTCTCGAATTCCTGACCTCAGGTGATCCACCCGCCCTGGCCTCCCAAAG";
    DnaString refRight = "AGAGACAGGGTTTTACCATGTTGGCCAGGCTGGTCTTGAGCTCCTGACCTCAGGTGATCCACCCGCCCTGGCCTCCCAAAGTGCTGGGATTACAGGCCTGAGCCGCTGCACCCGGCCCAGAGGCATTTATTTAAAATGTGCAATAAAAGTGTTCTGAGCTTCCCTTACGTACA";
    
    Gaps<DnaString> refGapsLeft;
    Gaps<DnaString> refGapsRight;
    Gaps<DnaString> readGapsLeft;
    Gaps<DnaString> readGapsRight;

    setSource(refGapsLeft, refLeft);
    setSource(refGapsRight, refRight);
    setSource(readGapsLeft, read);
    setSource(readGapsRight, read);

    Score<int> scoring(1, -3, -4, -5);
    
    AlignConfig<false, true, true, false> config;

    int splitScore = splitAlignment(readGapsLeft, refGapsLeft, readGapsRight, refGapsRight, scoring, config);

    std::cout << "refLeft:  " << refGapsLeft << "\n";
    std::cout << "readLeft: " << readGapsLeft << "\n\n";

    std::cout << "refRight:  " << refGapsRight << "\n";
    std::cout << "readRight: " << readGapsRight << "\n\n";

    std::cout << "Split score = " << splitScore;
```

Do I use the correct AlignConfig configuration?

**Comment by bkehr** [comment]
On another note, would it be possible to return a pair of scores in the split alignment function, i.e. the two scores of the left and right alignment, instead of only their sum?

**Comment by rrahn** [comment]
Ok it seems, the issue lies in the configuration of the AlignConfig object.
The way you want to align the left ref with the read, requires an overlap at the end of the read, since a semi-global alignment doesn't work well, if the sequences have equal length.
Hence with ```AlignConfig<false, true, true, true> config;``` you'll get your expected alignment.

**Comment by bkehr** [comment]
Excellent, it seems to work. I still need to test on larger input.

Apparently I misinterpreted the AlignConfig object:
I was assuming the first two booleans are for the left end of the left alignment and the last two booleans are for the right end of the right alignment. Then, the right end of the left alignment and the left end of the right alignment would be set internally by the splitAlignment function.

Do I understand correctly, that currently only symmetric configurations are allowed?

**Comment by rrahn** [comment]
Ah, no, apparently the AlignConfig is used for both left and right alignment. 
It might be necessary to add a second config to allow different parameterisation for left and right alignments. Do you need that? 
Second, do you urgently need the pair of scores instead of the sum. It is quite easily done, but it breaks the API, unless I add another interface that returns a pair instead of the sum, but it blows up the split interfaces a lot. 
My solution would be to adapt the impl interface to return a pair and the public interfaces return the sum.
Than you could use the impl version, which does not break the API.

**Comment by bkehr** [comment]
I don't currently need different configurations, I simply assumed you implemented the most general case.

I would prefer a pair of scores instead of the sum, yes.
What do you mean by public and impl interface?

**Comment by rrahn** [comment]
There is the standard public interface which you use at the moment.
All of these interfaces call this function: https://github.com/seqan/seqan/blob/master/include/seqan/align_split/align_split_interface.h#L287
I can change this internal function to return a tuple of left and right score, while the public interfaces still return a sum. In this case I don't break the API, but can add your feature, if you are willing to use impl interface. But it requires you to specify all function arguments, for instance the lower and upper diagonal.
You must explicitly unset them with ```minValue<TScoreValue>()``` and ```maxValue<TScoreValue>()``` if you don't want to call a banded version.

**Comment by bkehr** [comment]
Ok, that would be a solution.

As for the AlignConfig, I don't think the interface should allow passing a second config. I assumed the one config is split into left and right internally.

**Comment by rrahn** [comment]
Here is the new interface how you could use it to get a tuple of scores:
https://github.com/rrahn/seqan/blob/feature/split_align_tuple/include/seqan/align_split/align_split_interface.h#L532
The sum can than be computed as ```std::get<0>(tmp) + std::get<1>(tmp)```

**Comment by bkehr** [comment]
Now it doesn't compile anymore. Here is the first of many error messages:

```
In file included from seqan/score.h:49:0,
                 from seqan/graph_align.h:45,
                 from seqan/align.h:55,
                 from seqan/align_split.h:44,
                 from test_split_alignment.cpp:1:
seqan/score/score_matrix_dyn.h:110:1: warning: elaborated-type-specifier for a scoped enum must not use the ‘class’ keyword [enabled by default]
 enum class AminoAcidScoreMatrixID : std::underlying_type_t<decltype(Find<impl::score::MatrixTags, ScoreSpecBlosum30>::VALUE)>
 ^
seqan/score/score_matrix_dyn.h:110:12: error: use of enum ‘AminoAcidScoreMatrixID’ without previous declaration
 enum class AminoAcidScoreMatrixID : std::underlying_type_t<decltype(Find<impl::score::MatrixTags, ScoreSpecBlosum30>::VALUE)>

...
```

**Comment by bkehr** [comment]
Before I updated to your latest SeqAn version, I found this example that isn't returning the expected output even tough I now use `AlignConfig<false, true, true, true>`:

```
    DnaString read     =             "ATTTTTTTTTTTTTAAATTCTAATATTATGAAGAA";
    DnaString refLeft  = "ACACCCGGCTATTTTTTTTTTTTTTTTTT";
    DnaString refRight =          "TTCTTCTTCATTTATCTGAAATTCTAATATGATGAAGAATCCTGCTT";
```

Output of my test program (see above):
```
refLeft:  ACACCCGGCTATTTTTTTTTTTTTTTTTT
readLeft: ---------------ATTTTTTTTTTTTT

refRight:  AAATTCTAATATGATGAAGAATCCTGCTT
readRight: AAATTCTAATATTATGAAGAA--------
```

Expected output:
```

refLeft:  ACACCCGGCTATTTTTTTTTTTTT
readLeft: ----------ATTTTTTTTTTTTT

refRight:  AAATTCTAATATGATGAAGAATCCTGCTT
readRight: AAATTCTAATATTATGAAGAA--------
```

**Comment by rrahn** [comment]
The ```underlying_type_t``` is a c++14 feature. Can this be the problem?
Otherwise, please report to me your compiler version and the set flags so I can investigate this. 

I will look into the other issue on Monday.

**Comment by bkehr** [comment]
Ok, my compiler doesn't support c++14:
g++ (GCC) 4.8.5 20150623 (Red Hat 4.8.5-4)

**Comment by rrahn** [comment]
Yes, SeqAn 2.1 only supports >= GCC 4.9.

**Comment by bkehr** [comment]
In the Changelog for 2.1.0 it says "requirements are now GCC ≥ 4.8".
Do I have to go back to SeqAn 2.0.2?

**Comment by h-2** [comment]
> In the Changelog for 2.1.0 it says "requirements are now GCC ≥ 4.8".

Hm, you are right, but this is a typo. SeqAn-2.1.0 already only supported GCC-4.9 as the homepage and also the README indicate: https://github.com/seqan/seqan/tree/seqan-v2.1.0

We will support gcc-4.9 and clang-3.5 through the entire 2.x series and likely also vs2015 (albeit with installed updates).

**Comment by h-2** [comment]
CMake should have warned you loudly about the compiler if it is too old btw.

**Comment by bkehr** [comment]
Well, my program compiles with SeqAn 2.1 on GCC 4.8 unless I use René's fork, even though I use quite a few modules (bam_io, seq_io, vcf_io, align, index, arg_parse, ...).

I don't use CMake, I include seqan in a Makefile.

**Comment by h-2** [comment]
> Well, my program compiles with SeqAn 2.1 on GCC 4.8 unless I use René's fork

Then you got lucky ;-) Unfortunately there are some serious issues with gcc-4.8's c++11 support so we had to require 4.9 for the release.

> I don't use CMake, I include seqan in a Makefile.

We will add checks to the source code, as well, so that you are warned in the future if your compiler is outdated. See #1739

**Comment by rrahn** [comment]
@bkehr I fixed the split alignment. It was computing the wrong traceback. Anyway, I have now some issues with Gustaf. 
@ktrappe Can you verify the following changes to be correct?
```
 ##contig=<ID=gi|1234567|,length=1148>
 #CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT
 gi|9632547|ref|NC_002077.1|	210	.	G	<INV>	.	PASS	SVTYPE=INV;END=280;SVLEN=70;BM=6;DP=1	.	1
-gi|9632547|ref|NC_002077.1|	421	.	G	<DEL>	.	PASS	SVTYPE=DEL;END=490;SVLEN=-69;BM=0;DP=1	.	1
+gi|9632547|ref|NC_002077.1|	420	.	G	<DEL>	.	PASS	SVTYPE=DEL;END=490;SVLEN=-70;BM=0;DP=1	.	1
 gi|9632547|ref|NC_002077.1|	909	.	A	<INS>	.	PASS	SVTYPE=INS;SVLEN=70;BM=1;DP=1	.	1
-gi|9632547|ref|NC_002077.1|	1052	.	A	<DEL>	.	PASS	SVTYPE=DEL;END=1120;SVLEN=-68;BM=2;DP=1	.	1
+gi|9632547|ref|NC_002077.1|	1050	.	A	<DEL>	.	PASS	SVTYPE=DEL;END=1120;SVLEN=-70;BM=2;DP=1	.	1
 gi|9632547|ref|NC_002077.1|	1193	.	C	<DEL>	.	PASS	SVTYPE=DEL;END=1262;SVLEN=-69;BM=3;DP=1	.	1
 gi|9632547|ref|NC_002077.1|	1468	BND_5_1	G	G[gi|9632547|ref|NC_002077.1|:1821[	.	PASS	SVTYPE=BND;EVENT=Trans5;BM=4;DP=3	.	1
 gi|9632547|ref|NC_002077.1|	1469	BND_5_2	G	]gi|9632547|ref|NC_002077.1|:1888]G	.	PASS	SVTYPE=BND;EVENT=Trans5;BM=4;DP=3	.	1
```
The lines with the ```-``` have been replaced with the lines with the ```+```.
It seems like the begin positions of the deletions are shifted a little bit. 
So I investigated the split calculation for the first deletion at position 421 (old).
The original code (before my changes) resulted in the following alignment. It cuts before the ```X```
```
Left:                                                                  x
AGATCGTGATCAAGGTGCCGAGCGACCTGGACGAGCACCTGCCGGGCATTTCTGACTCGTTTGTGAGCT-GCTGACCGTGGCCGAGAAGCTGCAGCGCGACTTCCTGGTCCAATGGCGCCGCGTGAGTAAGGCCCCGGAGG
AGATCGTGATCAAGGTGCCGAGCGACCTGGACGAGCACCTGCCGGGCATTTCTGACTCGTTTGTGAGCTGGGTGGCCGAG-------------------------------------------------------------

Right:                                                                 x
AGATCGTGATCAAGGTGCCGAGCGACCTGGACGAGCACCTGCCGGGCATTTCTGACTCGTTTGTGAG-CTGCTGACCGTGGCCGAGAAGCTGCAGCGCGACTTCCTGGTCCAATGGCGCCGCGTGAGTAAGGCCCCGGAGG
-------------------------------------------------------------------ACCCCTGACCGTGGCCGAGAAGCTGCAGCGCGACTTCCTGGTCCAATGGCGCCGCGTGAGTAAGGCCCCGGAGG
```
This results in the following anchors:
```
Left:
AGATCGTGATCAAGGTGCCGAGCGACCTGGACGAGCACCTGCCGGGCATTTCTGACTCGTTTGTGAGCT-G
AGATCGTGATCAAGGTGCCGAGCGACCTGGACGAGCACCTGCCGGGCATTTCTGACTCGTTTGTGAGCTGG
Right:
CTGACCGTGGCCGAGAAGCTGCAGCGCGACTTCCTGGTCCAATGGCGCCGCGTGAGTAAGGCCCCGGAGG
CTGACCGTGGCCGAGAAGCTGCAGCGCGACTTCCTGGTCCAATGGCGCCGCGTGAGTAAGGCCCCGGAGG
```

We get this alignment, because in the original code we start the traceback from the last row or column of the dp matrix where we found the "global" maximum. This might be ok for many cases, but in such ambiguous paths, we might end up with the wrong one.
In the fixed version I compute the trace from the calculated split position.
This gives the following alignment:
```
Left:
AGATCGTGATCAAGGTGCCGAGCGACCTGGACGAGCACCTGCCGGGCATTTCTGACTCGTTTGTGAGCTG
AGATCGTGATCAAGGTGCCGAGCGACCTGGACGAGCACCTGCCGGGCATTTCTGACTCGTTTGTGAGCTG
Right:
CTGACCGTGGCCGAGAAGCTGCAGCGCGACTTCCTGGTCCAATGGCGCCGCGTGAGTAAGGCCCCGGAGG
CTGACCGTGGCCGAGAAGCTGCAGCGCGACTTCCTGGTCCAATGGCGCCGCGTGAGTAAGGCCCCGGAGG
```
Because it prefers the diagonal value in the cell of the split position, while the old code prefers left gap alignment exactly at the position of the breakpoint.
So my question would be, whether you can verify the correctness of the differences based on your validation process? I mean can the original deletion be recalled with the changed positions?

See the PR  #1741.

**Comment by ktrappe** [comment]
Hey, @bkehr is right, the traceback should start at the split-position. So the new version should be the formally correct one. This seems to be my toy example and the new breakpoints look also much more likely/"correct". I remember this to be an old issue. Long story short, I am fine with this.

**Comment by rrahn** [comment]
Ok, than I will adapt the gustaf tests and update the PR.
Thanks.

**Comment by bkehr** [comment]
I found another small issue: The function `_splitAlignmentImpl` now returns a tuple of integer references (if I am correct). This works in the small test program, but not my larger test cases, where the referenced values seem to be overwritten after returning from the function.

Replacing the last line of the function `return *res;` by the following works in my case:
```
std::tuple<int, int> ret = *res;
return ret;
```

Maybe there is a more elegant way of fixing this?

**Comment by rrahn** [comment]
No, it is fixed already. It returns now a std::pair of integers, but can be accessed in the same way as a tuple. 
Or you use ```first``` and ```second``` member.

**Comment by rrahn** [comment]
This merged PR #1743 should now have all features included, so I close this issue. 
@bkehr please reopen any time if there are more unattended use cases 😉

**Comment by bkehr** [comment]
Ah, I hadn't pulled your latest changes.
Thanks for fixing this!
